lib/Conversion/CalyxToFSM/MaterializeFSM.cpp

```diff
@@ -0,0 +1,199 @@
+//===- MaterializeCalyxToFSM.cpp - FSM Materialization Pass -----*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Contains the definitions of the FSM materialization pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "../PassDetail.h"
+#include "circt/Conversion/CalyxToFSM.h"
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/FSM/FSMGraph.h"
+#include "circt/Support/LLVM.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/OperationSupport.h"
+#include "llvm/ADT/STLExtras.h"
+
+using namespace circt;
+using namespace calyx;
+using namespace mlir;
+using namespace fsm;
+
+namespace {
+
+// A set of group names. Use SetVector to ensure deterministic ordering.
+using GroupSet = SetVector<StringAttr>;
+
+struct MaterializeCalyxToFSMPass
+    : public MaterializeCalyxToFSMBase<MaterializeCalyxToFSMPass> {
+  void runOnOperation() override;
+
+  struct StateEnableInfo {
+    llvm::SmallVector<Value> outputOperands;
+    llvm::SmallVector<Value> doneGuards;
+  };
+
+  /// Returns an operand vector for the states which should be enabled for a
```

Comment:
What is the consensus on using `///` versus `//` in LLVM?

Reply:
[`///` where doxygen may pay the code a visit](https://llvm.org/docs/CodingStandards.html#id13).  
I usually just do `///` to preface functions and member variables, and `//` inside functions.